sup3r/models/base.py

```diff
@@ -1052,3 +897,279 @@ def train(
+        train_gen,
+        train_disc,
+        disc_loss_bounds,
+        loss_mean_window=None,
+        multi_gpu=False,
+    ):
+        """Train the GAN for one epoch.
+
+        Parameters
+        ----------
+        batch_handler : sup3r.preprocessing.BatchHandler
+            BatchHandler object to iterate through
+        weight_gen_advers : float
+            Weight factor for the adversarial loss component of the generator
+            vs. the discriminator.
+        train_gen : bool
+            Flag whether to train the generator for this set of epochs
+        train_disc : bool
+            Flag whether to train the discriminator for this set of epochs
+        disc_loss_bounds : tuple
+            Lower and upper bounds for the discriminator loss outside of which
+            the discriminators will not train unless train_disc=True or
+            and train_gen=False.
+        loss_mean_window : int
+            Number of batches to use to compute generator and discriminator
+            loss means, which are used to decide whether to train each network
+            for a given batch. Defaults to the number of batches in an epoch
+        multi_gpu : bool
+            Flag to break up the batch for parallel gradient descent
+            calculations on multiple gpus. If True and multiple GPUs are
+            present, each batch from the batch_handler will be divided up
+            between the GPUs and resulting gradients from each GPU will be
+            summed and then applied once per batch at the nominal learning
+            rate that the model and optimizer were initialized with.
+            If true and multiple gpus are found, ``default_device`` device
+            should be set to /gpu:0
+
+        Returns
+        -------
+        loss_details : dict
+            Namespace of the breakdown of loss components
+        """
+
+        disc_th_low = np.min(disc_loss_bounds)
+        disc_th_high = np.max(disc_loss_bounds)
+        loss_means = (
+            {} if self._history.empty else self._history.iloc[-1].to_dict()
+        )
+        loss_means.setdefault('train_loss_disc', 0)
+        loss_means.setdefault('train_loss_gen', 0)
+        loss_mean_window = (
+            len(batch_handler)
+            if loss_mean_window is None
+            else loss_mean_window
+        )
+
+        only_gen = train_gen and not train_disc
+        only_disc = train_disc and not train_gen
+
+        if self._write_tb_profile:
+            tf.summary.trace_on(graph=True, profiler=True)
+
+        for ib, batch in enumerate(batch_handler):
+            b_loss_details = {}
+            loss_disc = loss_means['train_loss_disc']
+            disc_too_good = loss_disc <= disc_th_low
+            disc_too_bad = (loss_disc > disc_th_high) and train_disc
+            gen_too_good = disc_too_bad
+
+            if not self.generator_weights:
+                self.init_weights(batch.low_res.shape, batch.high_res.shape)
```

Comment:
Minor simplification - let's move this function call to the start of train() and put the `if not self.generator_weights` check in init_weights()